core/engine.go

```diff
@@ -219,6 +219,14 @@ type Diagnosable interface {
 	Diagnostics() []DiagnosticResult
 }
 
+// MappedDiagnosable allows the implementer to return a map of diagnostics providers,
+// which can be used by the caller to retrieve specific diagnostics, rather than everything (like Diagnosable provides).
+// The map keys should be URL paths, to aid mapping to HTTP APIs.
+type MappedDiagnosable interface {
```

Comment:
This adds a lot of complexity, do we expect any other use cases?

Reply:
I considered just registering a router and calling a separate diagnostics func for addressbook, but that leads to this special addressbook diagnostic func being propagated through all layers (Network -> ConnectionManager -> Addressbook), although Network doesn't care about gRPC impl's addressbook. So an abstracted concept is cleaner.

Reply:
fair.

Reply:
This still doesn't feel right. It moves routing decisions (url paths) to the diagnostics collectors. To me that sounds just as bad as the other solution. Maybe @woutslakhorst has any ideas?

Reply:
A design-wise cleaner solution would involve a Visitor-esque pattern (like API's `Routes()`) so components can register it themselves, e.g.:

```go
type DiagnosticProvider interface {
    Register(name string, provider func() []DiagnosticResult)
}
```

Which are then called by the `Status` engine and exposed as HTTP endpoints.

But currently we just have 1 separate diagnostics endpoint and it's unclear whether when, or even if, there will be more. So the added abstractions and indirections only add complexity which don't add any value i.m.o.

So yes, I agree from a design point of view, but it doesn't add any actual value right now.